server/src/main/java/org/opensearch/indices/pollingingest/DefaultStreamPoller.java

```diff
@@ -190,29 +192,27 @@ protected void startPoll() {
                 }
 
                 state = State.POLLING;
-
                 List<IngestionShardConsumer.ReadResult<? extends IngestionShardPointer, ? extends Message>> results;
 
-                if (includeBatchStartPointer) {
+                // todo: handle multi-writer scenarios to provide atleast once semantics
+                if (encounteredError && lastProcessedPointer != null) {
+                    results = consumer.readNext(lastProcessedPointer, false, MAX_POLL_SIZE, POLL_TIMEOUT);
+                } else if (includeBatchStartPointer) {
                     results = consumer.readNext(batchStartPointer, true, MAX_POLL_SIZE, POLL_TIMEOUT);
                 } else {
                     results = consumer.readNext(MAX_POLL_SIZE, POLL_TIMEOUT);
                 }
 
+                encounteredError = false;
                 if (results.isEmpty()) {
```

Comment:
i wonder in the corner case that the last processed message in the last batch runs into err after line 215, and also there's no new message after this message, then will this logic result in rewinding to the previous batch start pointer?

because this condition `results.isEmpty()` becomes true, and in the next iteration, `encounteredError ` becomes false, then it will assume the `lastProcessedPointer ` did not run ito error?